### ISSUE TITLE
Billion laughs vulnerability in group references

### ISSUE THREAD

**Issue by cchalm** [brunato/xmlschema#59]
This library is vulnerable to a variant of the [billion laughs attack (wikipedia link)](https://en.wikipedia.org/wiki/Billion_laughs_attack). Group references are expanded during parsing, which allows XSD parse time to increase exponentially with file size for specifically-crafted files. For example:

```XML
<?xml version="1.0" encoding="utf-8"?>

<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:group name="LOL1">
      <xs:sequence>
          <xs:element name="lol1" />
          <xs:element name="lol2" />
      </xs:sequence>
  </xs:group>

  <xs:group name="LOL2">
      <xs:sequence>
          <xs:group ref="LOL1" />
          <xs:group ref="LOL1" />
      </xs:sequence>
  </xs:group>

  <xs:group name="LOL3">
      <xs:sequence>
          <xs:group ref="LOL2" />
          <xs:group ref="LOL2" />
      </xs:sequence>
  </xs:group>

  ...etc...
</xs:schema>
```

If you continue this pattern for a total of 32 group definitions, parsing the XSD will take on the order of 100 years.

Hosted code that parses arbitrary XSDs using this library is vulnerable to DoS-type attacks. Malicious actors can upload XSDs crafted to exhaust the processing power of host machines.

Data from a base 10 variant (ten references per group) of the above XSD template.

| Number of groups | Parse Time (s) | Peak Memory Usage (B) |
|----------------------|-------------|---------|
| 1                             | 0.0082407  | 17736 |
| 2                             | 0.015668    | 17792 |
| 3                             | 0.022879    | 17824 |
| 4                             | 0.073479    | 17836 |
| 5                             | 0.62114      | 17860 |
| 6                             | 5.8897        | 17872 |
| 7                             | 60.61          | 17892 |
| 8                             | 601.33        | 17904 |

Memory use does not expand exponentially, making this less severe than the most common billion laughs vulnerability. However, if there are any optional allocations in the hot path, exponential memory use may be possible.

These are the calls that are made an exponential number of times. Built-in functions are not included. Not in call stack order:

| Filename:Lineno(Function) | Number of calls w/ 6 groups |
|---|---|
groups.py:319(iter_elements) | 6,666,731
groups.py:95(\_\_getitem\_\_) | 5,311,141
\_weakrefset.py:70(\_\_contains\_\_) | 4,444,505
\_collections_abc.py:630(\_\_iter\_\_) | 4,076,553
elements.py:263(ref) | 2,222,916
abc.py:178(\_\_instancecheck\_\_) | 2,222,249
elements.py:302(iter\_components)	| 2,222,220
groups.py:234(iter\_components) | 960,218
groups.py:106(\_\_len\_\_) | 124,334
schema.py:474(iter\_components) | 123,458

This vulnerability can be avoided by expanding group references during validation instead of during parsing, the same way type references are handled. When expanding during validation, number of expansions is linearly proportional to the size of the XML file being validated.

**Comment by brunato** [comment]
Thank you very much for the analysis. 

I'm working to a solution for this, that avoid groups expansion (essentially for a group reference the global group will be not copied entirely but only referenced, that means `self.extend(xsd_group)` will be changed to a `self.append(xsd_group)`). Generators for validating/decoding should already works with this (anyway I'm pretty confident that errors should be simple to fix) and the number of groups will be dramatically reduced.

Almost surely I will remove (or transform to a lazy property) the schema's `base_elements`, that are essentially the global elements plus the elements produced by global groups expansion. The building of this dictionary, that is not a necessary for validation, explodes with schemas with many nested groups (iter_elements).

An update for this will follow soon.

**Comment by galaxy4public** [comment]
I am not sure that the fix should be in this library since the problem is in the underlying libraries (in this case in the ElementTree).  It is addressed perfectly by using [defusedxml](https://pypi.org/project/defusedxml/), so why would @brunato spend his time on fixing something that is not part of the library?

P.S. While I am at it, I would love to thank @brunato for keeping up this work, it the most lightweight validator for Python I found so far and i love it :)

**Comment by cchalm** [comment]
Defusedxml looks like a great resource, I hadn't seen that before. However, this particular vulnerability comes from this library's non-symbolic representation of XSD group references; it is specific to the implementation of this library. Note that the vulnerability is not a standard XML bomb, which would be ElementTree's responsibility; it is an XSD-specific group reference expansion bomb.

**Comment by brunato** [comment]
I had already decided to use *defusedxml* into *xmlschema* from the next release, after a discussion with the author, recently at PyCon in Florence. The usage will be regulated by an optional argument, leaving open for a different choice to the users (there are schemas, for whatever reason, that contain entity definitions).

This vulnerability however is related to XSD groups representation, like  @cchalm wrote. This seems to be fixed in my develop, simply avoiding groups expansion (one more level for each group reference but no expansion) and counting on generators flexibility, but I've to check some test before considering the new version ready to be published.

**Comment by brunato** [comment]
A fix for this issue should be the one implemented within release 0.9.27, published yesterday.

**Comment by cchalm** [comment]
Verified fixed in 0.9.27. Thanks for the quick turnaround!
